src/ir/LocalGraph.cpp

```diff
@@ -103,12 +99,45 @@ struct Flower : public CFGWalker<Flower, Visitor<Flower>, Info> {
   }
 
   void flow(Function* func) {
+    // Minimalist block information needed
+    struct FlowBlock {
+      size_t lastTraversedIteration; // last Traversed Iteration
+      std::vector<Action> actions; // actions occurring in this block
+      std::vector<FlowBlock*> in;
+      std::vector<std::pair<Index, SetLocal*>> lastSets; // for each index, the last set_local for it
+    };
+
     auto numLocals = func->getNumLocals();
     std::vector<std::vector<GetLocal*>> allGets;
     allGets.resize(numLocals);
-    std::unordered_set<BasicBlock*> seen;
-    std::vector<BasicBlock*> work;
-    for (auto& block : basicBlocks) {
+    std::vector<FlowBlock*> work;
+
+    std::vector<FlowBlock> optimizedBlocks;
+    optimizedBlocks.resize(basicBlocks.size());
+
+    FlowBlock* entryFlowBlock = nullptr;
+    for (size_t i = 0; i < optimizedBlocks.size(); ++i) {
+      auto& optBlock = optimizedBlocks[i];
+      auto& inBlock = basicBlocks[i];
+      if (inBlock.get() == entry) entryFlowBlock = &optBlock;
+      optBlock.lastTraversedIteration = -1;
+      optBlock.actions.swap(inBlock->contents.actions);
+      auto& inBlocks = inBlock->in;
+      optBlock.in.reserve(inBlocks.size());
+      for (auto * pred : inBlocks) {
+        auto it = std::find_if(basicBlocks.begin(), basicBlocks.end(), [&](const std::unique_ptr<BasicBlock>& ptr) { return pred == ptr.get(); });
```

Comment:
if I understand correctly, there is a mapping between basic blocks and flow blocks, and this find_if here is finding the proper basic block for the flow block? this is linear in the number of blocks, how about instead creating a mapping once up front, then we just have a hashtable lookup here?